src/libraries/System.Data.Common/src/System/Xml/BaseTreeIterator.cs

```diff
@@ -49,7 +46,7 @@ internal bool NextRightRowElement()
         // Returns true if the current node is on a row element (head of a region)
```

Comment:
Line 16 of this file: CurrentNode should probably be nullable. This is an internal class, and is the base class for other two classes, which are in this same PR.
At some point, the underlying private field for the currentNode is set to null. For example, line 61 of `RegionIterator.cs`.

Reply:
Line 91 in `RegionIterator.cs` is doing the same, setting the value to null.

Reply:
`TreeIterator.cs` seems to be doing the same.

Reply:
Good point. Since this follows the typical iterator pattern, I have annotated the `NextXxx` methods with `[MemberNotNullIf(true, CurrentNode)]`. Sadly this doesn't do much for us because the compiler still can't infer that `iter.CurrentNode` isn't `null`. for cases like this:

```
for (bool fMore = iter.NextRowElement(); fMore; fMore = iter.NextRowElement())
{
    rowElem = (XmlBoundElement)(iter.CurrentNode!);
    EnsureDisconnectedDataRow(rowElem);
}
```